linux-firmware: cleanup around Cypress firmware and NVRAM packages

kuanyili opened the pull request:
This patchset performs some cleanup and refactoring around Cypress firmware and NVRAM packages. I'm not sure what the proper PR title prefix should look like. Please let me know if there is anything needs fixing.

In OpenWrt, Cypress WLAN is used primarily by Raspberry Pi boards, so it would be nice if @Noltari can have a look.

`cypress-nvram` is no longer used by Raspberry Pi boards after the cleanup, but PICO-PI-IMX7D still relies on this package, so keep it around for now.

Compile tested: bcm2711/RPi4B
Run tested: bcm2711/RPi4B

## References

- https://github.com/openwrt/openwrt/commit/922b538cc955a0dd3a7a381a116fba3983c2bc9e#commitcomment-69015650
- https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/log/brcm/brcmfmac43455-sdio.raspberrypi,3-model-b-plus.txt
- https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/log/brcm/brcmfmac43455-sdio.raspberrypi,4-model-b.txt
- https://github.com/openwrt/openwrt/commit/6f6c2fb32176b538a8d13d896e7842ae27950925#commitcomment-86915698

pepe2k commented:
CC: @Leo-PL

Leo-PL commented:
@pepe2k @kuanyili I wonder why we go towards Broadcom naming, since Cypress has bought Broadcom's Wi-Fi business, and not the other way around, and IIRC kernel starts to expect the cypress prefix for firmwares.
Anyway I'll try testing on pico-imx7d next week.

kuanyili commented:
@Leo-PL Ha, I know. That business unit went through Broadcom, Cypress, and now it's under Infineon, naming things around it could be a pain.

Have a look at the `WHENCE` file here
https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/WHENCE?h=20220913#n2727
to see how upstream `linux-firmware` organize files and links for driver `brcmfmac`.

Hope this explains a bit regarding the design choice.

Leo-PL commented:
@kuanyili Okay, now I understand. I plan to test your PR, but i highly doubt I can do that before sunday, I'm co-organizing a demoscene party, so I'll be very busy this weekend.

kuanyili commented:
@Leo-PL No worries! Thanks for your time. :)

kuanyili commented:
It turns out that `cypress-nvram` need some changes so SDK won't spit out errors below.

```
tmp/.config-package.in:9215:error: recursive dependency detected!
tmp/.config-package.in:9215:    symbol PACKAGE_brcmfmac-firmware-43430-sdio-rpi-3b is selected by PACKAGE_cypress-nvram-43430-sdio-rpi-3b
Config-build.in:6109:   symbol PACKAGE_cypress-nvram-43430-sdio-rpi-3b depends on PACKAGE_brcmfmac-firmware-43430-sdio-rpi-3b
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"

Config-build.in:6089:error: recursive dependency detected!
Config-build.in:6089:   symbol PACKAGE_brcmfmac-firmware-43455-sdio-rpi-4b is selected by PACKAGE_cypress-nvram-43455-sdio-rpi-4b
Config-build.in:6121:   symbol PACKAGE_cypress-nvram-43455-sdio-rpi-4b depends on PACKAGE_brcmfmac-firmware-43455-sdio-rpi-4b
For a resolution refer to Documentation/kbuild/kconfig-language.rst
subsection "Kconfig recursive dependency limitations"
```

Although packages still build, it is always nice to have them ironed out. I've added some commits at the beginning of the patchset to address this issue.

Leo-PL commented:
Works for me, even after stressing the wireless interface for 10 minutes with iperf3 in bidirectional mode, so:
`Tested-by: Lech Perczak <lech.perczak@gmail.com>`

hauke commented on package/firmware/cypress-nvram/Makefile:
```diff
@@ -53,14 +53,14 @@ $(eval $(call BuildPackage,cypress-nvram-43430-sdio-rpi-3b))
-	$(INSTALL_DATA) \
-		$(PKG_BUILD_DIR)/brcmfmac43430-sdio.raspberrypi,model-zero-w.txt \
+	$(LN) \
+		brcmfmac43430-sdio.raspberrypi,3-model-b.txt \
```
I would prefer if you merge all the nvram files at least for one wifi card together into one package. You could create a `cypress-nvram-43430` and add there all nvram files and all symlinks for 43430. They are pretty small and only text files. the ipkg should still be small.

Noltari replied:
@hauke @kuanyili I don't like this approach, since we can end up with a considerable amount of files for those devices with the most common SoCs, but I can live with that...

kuanyili replied:
@Noltari @hauke Those text files really are quite small. The combined text file size in `brcm` of `linux-firmware` is less than 36 kB. If we count only those for 43430, less than 4 kB; for 43455, less than 10 kB. And it will be even smaller after compression (SquashFS).

That's why I also pulled NVRAM files for other devices in while performing NVRAM package consolidation. This could benefit the support for many devices, including some RPi Zero 2 W boards. 

I'm fine with both though. If the decision is to keep NVRAM packages separate by device, just strip the last two commits away, and we are good to go.

Noltari replied:
@kuanyili I've just merged the first 6 commits, excluding the last two ones until we agree on what to do:
https://github.com/openwrt/openwrt/compare/ebe2b7190b7d8815a588eaf8a5cfdf9edfd85c36...a110418027a4c6f97c33980b8e4d7f07d39726de

Thanks!

kuanyili committed:
linux-firmware: broadcom: consolidate NVRAM packages

NVRAM packages for the same wireless chip are consolidated into one as
they contain only small text files and symlinks.

Signed-off-by: Kuan-Yi Li <kyli@abysm.org>

kuanyili committed:
cypress-nvram: consolidate NVRAM packages

NVRAM packages for the same wireless chip are consolidated into one as
they contain only small text files and symlinks.

Signed-off-by: Kuan-Yi Li <kyli@abysm.org>

kuanyili commented:
@hauke Done!

I've
- appended two additional commits to consolidate NVRAM packages
- rebased to latest master
- added proper release bumps accordingly (quite many of them as I tried to do it incrementally and not to break things between commits)

Compile tested: bcm27xx/bcm2711, imx/cortexa7
Run tested: bcm2711/RPi4B

Noltari commented:
@kuanyili excuse me for not looking into this before...

> This patchset performs some cleanup and refactoring around Cypress firmware and NVRAM packages. I'm not sure what the proper PR title prefix should look like. Please let me know if there is anything needs fixing.

Broadcom/Cypress/Infineon WLAN firmwares always involve headaches, so I totally understand you :)

> In OpenWrt, Cypress WLAN is used primarily by Raspberry Pi boards, so it would be nice if @Noltari can have a look.

Yes, I introduced that package when Cypress acquired that part from Broadcom.
The firmware files needed for the newer Raspberry Pi models weren't available on any repository with a proper LICENSE, and the only place were one could get them was directly from Cypress.
This usually happens whenever a new Raspberry Pi model is introduced.

> `cypress-nvram` is no longer used by Raspberry Pi boards after the cleanup, but PICO-PI-IMX7D still relies on this package, so keep it around for now.

Actually, commit 2e297377bf77f6051e5fed8c4fb684fd17d41987 is wrong since it should have added brcmfmac4339-sdio.technexion,imx7d-pico-pi.txt at https://github.com/openwrt/cypress-nvram and not in the files folder of the OpenWrt package...
Anyway, as I said before, it should probably be wise to keep cypress-nvram because when new devices are introduced we need additional files for the nvram which usually take a lot of time to be upstreamed to linux-firmware...

Should we get rid of cypress-firmware now that all files have been moved to linux-firmware?
Also, there's now https://github.com/Infineon/ifx-linux-firmware/tree/master/firmware which sould replace cypress-firmware, right?

kuanyili commented:
@Noltari Hey! Thanks for looking into this.

> Actually, commit [2e29737](https://github.com/openwrt/openwrt/commit/2e297377bf77f6051e5fed8c4fb684fd17d41987) is wrong since it should have added brcmfmac4339-sdio.technexion,imx7d-pico-pi.txt at https://github.com/openwrt/cypress-nvram and not in the files folder of the OpenWrt package... Anyway, as I said before, it should probably be wise to keep cypress-nvram because when new devices are introduced we need additional files for the nvram which usually take a lot of time to be upstreamed to linux-firmware...

If you ask me, I would probably

- keep `cypress-nvram` package for future NVRAM that is pending to be upstreamed
- disassociate `cypress-nvram` package from GitHub https://github.com/openwrt/cypress-nvram repository and keep NVRAM files here like `ipq-wifi` package did because
    - NVRAM files are even smaller than board files
    - it lowers maintenance cost
    - device support can be implemented here all in one go, instead of upstreaming NVRAM file first to https://github.com/openwrt/cypress-nvram

> Should we get rid of cypress-firmware now that all files have been moved to linux-firmware? Also, there's now https://github.com/Infineon/ifx-linux-firmware/tree/master/firmware which sould replace cypress-firmware, right?

I'm inclined to keep `cypress-firmware` package which already utilizes https://github.com/Infineon/ifx-linux-firmware as source. Firmware from there tend to be more up-to-date than those from `linux-firmware`. In the future, vulnerabilities similar to Kr00k and FragAttacks might got fixed faster there.

We can then switch between `brcmfmac-firmware-xxxx` and `cypress-firmware-xxxx` whenever needed.

kuanyili committed:
cypress-nvram: disassociate from external source repo

Since all NVRAM files in external repo are now upstreamed and to lower
future maintenance cost, disassociate the package from external source
repo.

All upstream pending NVRAM files shall be stored locally from now on.

Signed-off-by: Kuan-Yi Li <kyli@abysm.org>

kuanyili commented:
Awesome! Thanks guys!

To tidy things up, the PR branch is rebased and inserted an external repo disassociation commit.

So there are now

- 1 disassociation commit
- 2 consolidation commit (up for discussion)

Compile tested: bcm27xx/bcm2711
Run tested: bcm2711/RPi4B

Noltari commented:
> Awesome! Thanks guys!
> 
> To tidy things up, the PR branch is rebased and inserted an external repo disassociation commit.
> 
> So there are now
> 
> * 1 disassociation commit
> * 2 consolidation commit (up for discussion)
> 
> Compile tested: bcm27xx/bcm2711 Run tested: bcm2711/RPi4B

Thanks!

I've made up my mind and I will also merge the consolidation commits later today :)

Noltari commented:
Merged, thanks! https://github.com/openwrt/openwrt/compare/f74275c936e9cc62959822746270abb60339d18b...49bbfd9968879f3aa6df07871703a2643254bf60
